File: python/day4.py
import re

with open("/home/patrick/Documents/projects/1problem-3languages/day4_input.txt", "r") as f:
    file_content = f.read()

def create_passport(l):
    kv_pairs = l.replace("\n", " ").split()
    passport = {}

    for p in kv_pairs:
        kv = p.split(":")
        passport[kv[0]] = kv[1]

    return passport

passports = list(map(lambda x: create_passport(x), file_content.split("\n\n")))


# PART 1
# missing cid is fine but missing any other is not

def is_valid_passport(p):
    required_fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"] # cid is optional

    if "cid" in p:
        del p["cid"]

    return (sorted(required_fields) == sorted(list(p.keys())))

print(len(list(filter(lambda x: is_valid_passport(x), passports))))

# PART 2

def is_valid_birth_year(p):
    if "byr" in p:
        return int(p["byr"]) >= 1920 and int(p["byr"]) <= 2002
    else:
        return False

def is_valid_issue_year(p):
    if "iyr" in p:
        return int(p["iyr"]) >= 2010 and int(p["iyr"]) <= 2020
    else:
        return False

def is_valid_exp_year(p):
    if "eyr" in p:
        return int(p["eyr"]) >= 2020 and int(p["eyr"]) <= 2030
    else:
        return False

def is_valid_height(p):
    if "hgt" in p:
        unit = re.findall("[a-zA-Z]+", p["hgt"])
        value = int(re.findall('[0-9]+', p["hgt"])[0])

        if len(unit) == 1 and unit[0] == "cm":
            return value >= 150 and value <= 193
        elif len(unit) == 1 and unit[0] == "in":
            return value >= 59 and value <= 76
        else:
            return False
    else:
        return False

def is_valid_hair(p):
    if "hcl" in p:
        color = re.findall("[a-zA-Z0-9]+", p["hcl"])[0]
        return p["hcl"][0] == "#" and len(color) == 6
    else:
        return False

def is_valid_eye(p):
    if "ecl" in p:
        return p["ecl"] in ["amb", "blu", "brn", "gry", "grn", "hzl", "oth"]
    else:
        return False

def is_valid_pid(p):
    if "pid" in p:
        pid = re.findall("[0-9]+", p["pid"])[0]
        return len(pid) == 9
    else:
        return False

def stricter_passport_check(p):
    return is_valid_birth_year(p)\
       and is_valid_issue_year(p)\
       and is_valid_exp_year(p)\
       and is_valid_height(p)\
       and is_valid_hair(p)\
       and is_valid_eye(p)\
       and is_valid_pid(p)

print(len(list(filter(lambda x: stricter_passport_check(x), passports))))

File: clojure/day4.clj
(ns aoc-2020.day4
  (:require
   [clojure.string :as string]
   [medley.core :as medley]))

(defn create-passport [s]
  (let [->kv-vec (fn [kv-str] (string/split kv-str #":"))
        kv-pairs (-> s
                     (string/replace "\n" " ")
                     (string/split #" "))]
    (->> kv-pairs
         (map ->kv-vec)
         (into {})
         (medley/map-keys keyword))))

(def passports (-> (slurp "/home/patrick/Documents/projects/1problem-3languages/day4_input.txt")
                   (string/split #"\n\n")
                   (#(map create-passport %))))

;; PART 1

(defn valid-passport? [p]
  (let [required-fields [:byr :iyr :eyr :hgt :hcl :ecl :pid]]
    (= (sort required-fields) (-> p
                                  (dissoc :cid)
                                  keys
                                  sort))))

(count (filter valid-passport? passports))
;; => 239

;; PART 2

(defn- read-string-when-not-empty [s]
  (when (not (string/blank? s))
    (read-string s)))

(defn valid-byr? [{byr :byr}]
  (when-let [year (some->> byr
                           (re-find #"\d+")
                           (read-string-when-not-empty))]
    (and (>= year 1920)
         (<= year 2002))))

(defn valid-iyr? [{iyr :iyr}]
  (when-let [year (some->> iyr
                           (re-find #"\d+")
                           (read-string-when-not-empty))]
    (and (>= year 2010)
         (<= year 2020))))

(defn valid-eyr? [{eyr :eyr}]
  (when-let [year (some->> eyr
                           (re-find #"\d+")
                           (read-string-when-not-empty))]
    (and (>= year 2020)
         (<= year 2030))))

(defn valid-hgt? [{hgt :hgt}]
  (when hgt
    (let [height (some->> hgt
                          (re-find #"\d+")
                          (read-string-when-not-empty))
          unit   (some->> hgt
                          (re-find #"\D+"))]
      (case unit
        "cm"
        (and (>= height 150) (<= height 193))

        "in"
        (and (>= height 59) (<= height 76))

        nil))))

(defn valid-hcl? [{hcl :hcl}]
  (when hcl
    (and (= 7 (count hcl))
         (= \# (first hcl))
         (= 6 (->> hcl (re-find #"[a-zA-Z0-9]+") (count))))))

(defn valid-ecl? [{ecl :ecl}]
  (when ecl
    (contains? #{"amb" "blu" "brn" "gry" "grn" "hzl" "oth"}
               (re-find #"[a-z]+" ecl))))

(defn valid-pid? [{pid :pid}]
  (when pid
    (and (every? #(java.lang.Character/isDigit %) (seq pid))
         (= 9 (count pid)))))

(defn is-valid-passport? [p]
  (and (valid-byr? p)
       (valid-iyr? p)
       (valid-eyr? p)
       (valid-hgt? p)
       (valid-hcl? p)
       (valid-ecl? p)
       (valid-pid? p)))

(count (filter is-valid-passport? passports))
;; => 188
